ADAL/src/cache/ios/ADKeychainTokenCache.m

```diff
@@ -245,30 +245,48 @@ - (ADTokenCacheItem*)itemFromKeyhainAttributes:(NSDictionary*)attrs
 
 
 /*!
-    @param  item    The item to remove from the cache
+    @param  item    The item to be set as tombstone in cache
     @param  error   (Optional) In the case of an error this will be filled with the
                     error details.
  
-    @return YES if the item was successfully removed, or was not in the cache. If NO
-            look
+    @return YES if the item was successfully tombstoned or not in the cache.
  */
 - (BOOL)removeItem:(nonnull ADTokenCacheItem *)item
              error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
 {
     RETURN_NO_ON_NIL_ARGUMENT(item);
     
+    //try to delete the old item
     ADTokenCacheKey* key = [item extractKey:error];
     if (!key)
     {
         return NO;
     }
-    
-    
     NSMutableDictionary* query = [self queryDictionaryForKey:key
                                                       userId:item.userInformation.userId
                                                   additional:nil];
     OSStatus status = SecItemDelete((CFDictionaryRef)query);
```

Comment:
Just do a SecItemUpdate instead. If that fails then we know the item wasn't in the cache in the first place.